File: backend/app/schemas/institution_knowledge.py
"""Pydantic schemas for the Institution Knowledge Foundation read-only API."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class _ORMModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class CampusRead(_ORMModel):
    id: uuid.UUID
    name: str
    city: str | None = None
    province: str | None = None
    is_main_campus: bool
    is_active: bool
    data_status: str
    is_synthetic: bool
    source_url: str | None = None


class ContactRead(_ORMModel):
    id: uuid.UUID
    name: str | None = None
    role: str | None = None
    email: str | None = None
    phone: str | None = None
    department: str | None = None
    contact_type: str | None = None
    is_public: bool
    data_status: str
    is_synthetic: bool


class AccreditationRead(_ORMModel):
    id: uuid.UUID
    body_id: uuid.UUID
    programme_id: uuid.UUID | None = None
    status: str
    accredited_date: date | None = None
    expiry_date: date | None = None
    notes: str | None = None
    data_status: str
    is_synthetic: bool


class PolicyRead(_ORMModel):
    id: uuid.UUID
    title: str
    policy_type: str | None = None
    description: str | None = None
    is_active: bool
    data_status: str
    is_synthetic: bool
    source_url: str | None = None


class InstitutionDocumentRead(_ORMModel):
    id: uuid.UUID
    title: str
    document_type: str | None = None
    description: str | None = None
    document_url: str | None = None
    publication_year: int | None = None
    data_status: str
    is_synthetic: bool
    source_url: str | None = None


class FacultyBrief(_ORMModel):
    id: uuid.UUID
    name: str
    code: str


class InstitutionProfile(BaseModel):
    """Full institution profile. Student callers receive a filtered view."""

    id: uuid.UUID
    name: str
    code: str
    country: str | None = None
    province: str | None = None
    website: str | None = None
    logo_url: str | None = None
    data_status: str | None = None
    is_demo: bool = False
    campuses: list[CampusRead] = []
    faculties: list[FacultyBrief] = []
    contacts: list[ContactRead] = []
    accreditations: list[AccreditationRead] = []


class ProvenanceBreakdown(BaseModel):
    public_verified: int = 0
    needs_review: int = 0
    synthetic_demo: int = 0
    customer_data: int = 0


class EntityCoverage(BaseModel):
    total: int = 0
    by_status: ProvenanceBreakdown = ProvenanceBreakdown()


class InstitutionCoverage(BaseModel):
    institution_id: uuid.UUID
    institution_code: str
    institution_name: str
    counts: dict[str, int]
    provenance: ProvenanceBreakdown
    ready_for_rag: bool
    generated_at: datetime


class KnowledgeOverview(BaseModel):
    institutions: int
    campuses: int
    faculties: int
    schools: int
    departments: int
    programmes: int
    qualifications: int
    modules: int
    learning_outcomes: int
    graduate_attributes: int
    policies: int
    institution_documents: int
    accreditation_bodies: int
    accreditations: int
    contacts: int
    provenance: ProvenanceBreakdown

File: frontend/src/lib/api/institutionKnowledge.ts
import { apiClient } from "@/lib/api-client";

export interface ProvenanceBreakdown {
  public_verified: number;
  needs_review: number;
  synthetic_demo: number;
  customer_data: number;
}

export interface InstitutionCoverage {
  institution_id: string;
  institution_code: string;
  institution_name: string;
  counts: Record<string, number>;
  provenance: ProvenanceBreakdown;
  ready_for_rag: boolean;
  generated_at: string;
}

export interface KnowledgeOverview {
  institutions: number;
  campuses: number;
  faculties: number;
  schools: number;
  departments: number;
  programmes: number;
  qualifications: number;
  modules: number;
  learning_outcomes: number;
  graduate_attributes: number;
  policies: number;
  institution_documents: number;
  accreditation_bodies: number;
  accreditations: number;
  contacts: number;
  provenance: ProvenanceBreakdown;
}

export interface CampusRead {
  id: string;
  name: string;
  city: string | null;
  province: string | null;
  is_main_campus: boolean;
  is_active: boolean;
  data_status: string;
  is_synthetic: boolean;
  source_url: string | null;
}

export interface ContactRead {
  id: string;
  name: string | null;
  role: string | null;
  email: string | null;
  phone: string | null;
  department: string | null;
  contact_type: string | null;
  is_public: boolean;
  data_status: string;
  is_synthetic: boolean;
}

export interface AccreditationRead {
  id: string;
  body_id: string;
  programme_id: string | null;
  status: string;
  accredited_date: string | null;
  expiry_date: string | null;
  notes: string | null;
  data_status: string;
  is_synthetic: boolean;
}

export interface InstitutionProfile {
  id: string;
  name: string;
  code: string;
  country: string | null;
  province: string | null;
  website: string | null;
  logo_url: string | null;
  data_status: string | null;
  is_demo: boolean;
  campuses: CampusRead[];
  faculties: { id: string; name: string; code: string }[];
  contacts: ContactRead[];
  accreditations: AccreditationRead[];
}

export interface LiveCountsResponse {
  institution_id: string | null;
  campuses: number;
  faculties: number;
  schools: number;
  departments: number;
  programmes: number;
  qualifications: number;
  modules: number;
  learning_outcomes: number;
  graduate_attributes: number;
  policies: number;
  policy_versions: number;
  institution_documents: number;
  accreditation_bodies: number;
  accreditations: number;
  contacts: number;
}

export interface CoverageSummaryResponse {
  institution_id: string | null;
  total_entities: number;
  public_verified_pct: number;
  needs_review_pct: number;
  synthetic_demo_pct: number;
  customer_data_pct: number;
  rag_readiness: "ready" | "partial" | "not_ready";
  crawler_readiness: "ready" | "partial" | "not_ready";
  missing_data_warnings: string[];
}

export interface FullInstitutionProfile {
  id: string;
  name: string;
  code: string;
  country: string | null;
  province: string | null;
  website: string | null;
  logo_url: string | null;
  data_status: string | null;
  is_demo: boolean;
  campuses: CampusRead[];
  faculties: { id: string; name: string; code: string; department_count: number }[];
  policies: {
    id: string;
    title: string;
    policy_type: string | null;
    description: string | null;
    is_active: boolean;
    data_status: string;
    is_synthetic: boolean;
    source_url: string | null;
  }[];
  documents: {
    id: string;
    title: string;
    document_type: string | null;
    description: string | null;
    document_url: string | null;
    publication_year: number | null;
    data_status: string;
    is_synthetic: boolean;
    source_url: string | null;
  }[];
  accreditations: AccreditationRead[];
  contacts: ContactRead[];
  coverage: Record<string, number>;
}

export async function getKnowledgeOverview(): Promise<KnowledgeOverview> {
  const { data } = await apiClient.get<KnowledgeOverview>(
    "/institution-knowledge/overview"
  );
  return data;
}

export async function getInstitutionCoverage(
  institutionId: string
): Promise<InstitutionCoverage> {
  const { data } = await apiClient.get<InstitutionCoverage>(
    `/institution-knowledge/institutions/${institutionId}/coverage`
  );
  return data;
}

export async function getInstitutionKnowledgeProfile(
  institutionId: string
): Promise<InstitutionProfile> {
  const { data } = await apiClient.get<InstitutionProfile>(
    `/institution-knowledge/institutions/${institutionId}/profile`
  );
  return data;
}

export async function getLiveCounts(
  institutionId?: string
): Promise<LiveCountsResponse> {
  const params = institutionId ? `?institution_id=${institutionId}` : "";
  const { data } = await apiClient.get<LiveCountsResponse>(
    `/institution-knowledge/live-counts${params}`
  );
  return data;
}

export async function getCoverageSummary(
  institutionId?: string
): Promise<CoverageSummaryResponse> {
  const params = institutionId ? `?institution_id=${institutionId}` : "";
  const { data } = await apiClient.get<CoverageSummaryResponse>(
    `/institution-knowledge/coverage-summary${params}`
  );
  return data;
}

export async function getFullInstitutionProfile(
  institutionId: string
): Promise<FullInstitutionProfile> {
  const { data } = await apiClient.get<FullInstitutionProfile>(
    `/institution-knowledge/institutions/${institutionId}/full-profile`
  );
  return data;
}
